File: #26-50 (Ada, C++, FORTRAN)/C++/Euler47.cpp
/////////////////////////////////////////////////////////////////////
// Euler47 ~ Distinct primes factors
// 
// Problem Statement:
// 
// The first two consecutive numbers to have two distinct prime
// factors are:
// 
//		14 = 2 × 7
//		15 = 3 × 5
// 
// The first three consecutive numbers to have three distinct prime
// factors are:
// 
//		644 = 2^2 × 7 × 23
//		645 = 3   × 5 × 43
//		646 = 2   × 17 × 19.
// 
// Find the first four consecutive integers to have four distinct
// prime factors each. What is the first of these numbers?
//
/////////////////////////////////////////////////////////////////////

/////////////////////////////////////////////////////////////////////
// Algorithm Explanation:
// 
// Iterates upwards getting the factors of each number and checking
// if there are at least four prime factors. Increments consecutive
// when found and resets to zero if the next numbers do not contain
// four prime factors. Once consecutive reaches four the program
// breaks.
//
/////////////////////////////////////////////////////////////////////

#include <iostream>
#include <math.h>
#include <chrono>

using namespace std;

bool prime(int num)
{
	if (num == 1) return 0;
	for (int i = 2; i <= sqrt(num); i++)
	{
		if (num % i == 0)
		{
			return 0;
		}
	}
	return 1;
}

int main()
{
	auto start = std::chrono::high_resolution_clock::now();
	int solution = 0;

	int p, consecutive = 0;
	for (int i = 646; i < 1000000; i++)
	{
		p = 0;
		for (int j = 1; j <= i/2; j++)
		{
			if (i % j == 0)
			{
				if (prime(j))
				{
					p++;
				}
			}
		}
		if (p >= 4)
		{
			consecutive++;
		}
		else
		{
			consecutive = 0;
		}
		if (consecutive == 4)
		{
			solution = i;
			break;
		}
	}

	solution = solution - 3; //we want the first of the four consecutive numbers

	auto done = std::chrono::high_resolution_clock::now();
	cout << "Solution: " << solution << endl << endl;
	cout << "Program completed in " << std::chrono::duration_cast<std::chrono::milliseconds>(done - start).count() << " milliseconds." << endl;
}

File: #26-50 (Ada, C++, FORTRAN)/FORTRAN/Euler47.f90
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
! Euler47 ~ Distinct primes factors
! 
! Problem Statement:
! 
! The first two consecutive numbers to have two distinct prime
! factors are:
! 
!		14 = 2 × 7
!		15 = 3 × 5
! 
! The first three consecutive numbers to have three distinct prime
! factors are:
! 
!		644 = 2^2 × 7 × 23
!		645 = 3   × 5 × 43
!		646 = 2   × 17 × 19.
! 
! Find the first four consecutive integers to have four distinct
! prime factors each. What is the first of these numbers?
!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
! Algorithm Explanation:
! 
! Iterates upwards getting the factors of each number and checking
! if there are at least four prime factors. Increments consecutive
! when found and resets to zero if the next numbers do not contain
! four prime factors. Once consecutive reaches four the program
! breaks.
!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

program Euler47
	
	implicit none
	integer :: time_start, time_finish, clock_finish, solution = 0
	integer :: i, j, p, consecutive = 0
	call SYSTEM_CLOCK(time_start)
	
	do i = 646,1000000
		p = 0
		do j = 1,(i/2)
			if (modulo(i,j) == 0) then
				if (prime(j)) then
					p = p + 1
				end if
			end if
		end do
		if (p >= 4) then
			consecutive = consecutive + 1
		else
			consecutive = 0
		end if
		if (consecutive == 4) then
			solution = i
			EXIT
		end if
	end do

	solution = solution - 3 !we want the first of the four consecutive numbers

	call SYSTEM_CLOCK(time_finish)
	clock_finish = time_finish - time_start
	write (*,"(A10,I10,//,A21,I5,A14)") "Solution: ",solution,"Program completed in ",clock_finish," milliseconds."

contains

	logical function prime(num)
		integer :: num, i
		if (num == 1) then
			prime = .false.
			return
		end if
		do i = 2,INT(sqrt(real(num)))
			if(modulo(num,i) == 0) then
				prime = .false.
				return
			end if
		end do
		prime = .true.
		return
    end

end program Euler47
